pyfifotax/report_data.py

```diff
@@ -0,0 +1,548 @@
+            ]
+        )
+
+    def process_report_events(self):
+        for event in self.report_events:
+            if isinstance(event, RSUEvent):
+                self.held_shares[event.symbol].push(event.received_shares)
+
+                bonus = awv.AWVEntryZ4RSUBonus(
+                    date=event.date,
+                    symbol=event.symbol,
+                    currency=event.received_shares.currency,
+                    value=event.received_shares.total_buy_value()
+                    + event.withheld_shares.total_buy_value(),
+                )
+                bought_shares = awv.AWVEntryZ10RSUDeposit(
+                    date=event.date,
+                    symbol=event.symbol,
+                    currency=event.received_shares.currency,
+                    quantity=event.received_shares.quantity
+                    + event.withheld_shares.quantity,
+                    value=event.received_shares.total_buy_value()
+                    + event.withheld_shares.total_buy_value(),
+                )
+                withheld_shares = awv.AWVEntryZ10RSUTaxWithholding(
+                    date=event.date,
+                    symbol=event.symbol,
+                    currency=event.received_shares.currency,
+                    quantity=event.withheld_shares.quantity,
+                    value=event.received_shares.total_buy_value(),
+                )
+                self.awv_z4_events.append(bonus)
+                self.awv_z10_events.append(bought_shares)
+                if event.withheld_shares.quantity > 0:
+                    self.awv_z10_events.append(withheld_shares)
+
+            elif isinstance(event, ESPPEvent):
+                self.held_shares[event.symbol].push(event.received_shares)
+
+                bonus = awv.AWVEntryZ4ESPPBonus(
+                    symbol=event.symbol,
+                    currency=event.currency,
+                    date=event.date,
+                    value=event.bonus,
+                )
+
+                bought_shares = awv.AWVEntryZ10ESPPDeposit(
+                    date=event.date,
+                    symbol=event.symbol,
+                    currency=event.currency,
+                    quantity=event.received_shares.quantity,
+                    value=event.received_shares.total_buy_value(),
+                )
+
+                self.awv_z4_events.append(bonus)
+                self.awv_z10_events.append(bought_shares)
+
+            elif isinstance(event, DividendEvent):
+                self.held_forex[event.currency].push(event.received_net_dividend)
+
+                if event.withheld_tax is not None:
+                    self.misc["Tax Withholding"].append(event.withheld_tax)
+                self.misc["Dividend Payments"].append(event.received_dividend)
+
+                # dividends should be small enough to not trigger AWV reportings
+
+            elif isinstance(event, BuyEvent):
+                # if not enough money, pop on FOREX Queue will fail
```

Comment:
I bumped into that some exchanges always calculate their fees in EUR. It seems this is the case with DEGIRO, and IBKR currency conversion is always deducted in EUR, too. However, IBKR stock exchange fees are in their original nomination (e.g. USD).

Even I feel here that adding a new column (fee currency) would be too much, but I don't know how widespread this practice is among different brokers. If it's really just DEGIRO, I figured out that I can model the EUR fees as a "0 € money withdrawal event with _x_ € fee".

So, it's not a problem in my case but if it's a widespread practice among multiple brokers, I'd add this new column. Perhaps we could ask around.

Reply:
There is a general issue with fees for transactions: technically they lower potential capital gains. So ideally, they should be considered together with these transactions to begin with.
As we initially only focused on Schwab, transactional fees usually are so low that they don't really matter and wire transfer fees e.g. can potentially be considered as Werbungskosten.

Reply:
I'd definitely avoid another column for fee currency but rather would recommend to just convert within the conversion scripts to the currency of received forex (internally, the same rate then should be used to convert back to EUR fees)

Reply:
In general, I find this kind of an odd behavior. This means that you always have to keep a certain balance of a specific currency only for fees, right?

Reply:
The issue with "generic" fees is that 
- technically, transactions costs can offset capital gains
- the most recent update now actually has this behavior
- if you just track fees generically in a different place, you loose that behavior

If that is ok (or if you manually consider this a later point), my recommendation would be to model generic fees as a "withdrawal" event
if you e.g. withdraw "FEE" amounts of a certain currency and set the fee amount also to "FEE", you have the same effect, i also updated the event so that the resulting fees will pass through the comment of the withdrawal event, so the list of fees should contain enough information

Reply:
> In general, I find this kind of an odd behavior. This means that you always have to keep a certain balance of a specific currency only for fees, right?

Yes, that's the EUR balance. I find it cleaner to think about: "I exchanged 1 EUR to 2 USD". Additional fee: 0.5 EUR. Instead of: "I exchanged 1 EUR to 1 USD with 1:2 conversion rate. What's wrong with the rate, then?".

Mathematically, you can move it back and forth, it's just I like it more because of the better separation and visibility of the market rates.

> if you e.g. withdraw "FEE" amounts of a certain currency and set the fee amount also to "FEE", you have the same effect

Just to confirm if I understand it right: if I have a 10 € account management fee, then I create the following in the money_transfers sheet:

date | buy_date | amount | fees | currency | comment
-- | -- | -- | -- | -- | --
2020-01-01 | 2020-01-01 | -10 | 10 | EUR | Account management fee [on IBKR]

Reply:
>Just to confirm if I understand it right: if I have a 10 € account management fee, then I create the following in the money_transfers sheet:

Yes, that should work. The fee for a withdrawal is always popped first and the remaining "net" amount is actually withdrawn. This should fit your use-cases with the least changes.

> Yes, that's the EUR balance. I find it cleaner to think about: "I exchanged 1 EUR to 2 USD". Additional fee: 0.5 EUR. Instead of: "I exchanged 1 EUR to 1 USD with 1:2 conversion rate. What's wrong with the rate, then?".

Technically, I agree. Although I am not a huge fan of separating things into too many transactions as one could forget a few quite easily.
- initially, you made it sound like a USD->AUD exchange e.g. also triggers EUR fees, I'd find this annoying as a user of a broker to be honest 
- for EUR/USD conversions, things basically then all evolve around the point of view of whether fees are inclusive or exclusive, a Schwab USD->EUR exchange for technically trigger fees at both ends, and while yes, 1 EUR to USD is the cleanest way, this doesn't tell you how the balances on both ends have moved, the "inclusive" way would be saying I exchanged 1.5 EUR to 1.5 USD with both a 0.50 EUR fee and a 0.50 USD fee (underlying then is the 1:2 exchange)
- although I admit that things are not really consistent anyways as e.g. selling shares comes with a separate fee

Seeing all these things play out, I could indeed imagine that a separate "fee currency" column in all sheets might be the most elegant solution.